### PR TITLE
Major data structures refactoring & enhancement

### PR DESCRIPTION
**Pull request by Secret333Boy**

### CODE REVIEW

File: src/tools/dataStructures/Matrix.js [hunk 1]
```diff
@@ -96,6 +162,14 @@ class Matrix {
   get cols() {
     return this.transpone().arr;
   }
+
+  set cols(arr) {
```
Comment (GrigoriyUstymenko) [inline src/tools/dataStructures/Matrix.js:166]: Input is not checked

File: src/tools/dataStructures/Matrix.js [hunk 2]
```diff
@@ -96,6 +162,14 @@ class Matrix {
...
+    this.arr = new Matrix(arr).transpone().arr;
+  }
+
+  set rows(arr) {
```
Comment (GrigoriyUstymenko) [inline src/tools/dataStructures/Matrix.js:170]: Input is not checked

File: src/tools/dataStructures/PriorityQueue.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        pointer.nextNode &&
+        pointer.nextNode.priority * this.direction >= priority * this.direction
+      ) {
+        pointer = pointer.nextNode;
+      }
+
+      const buf = pointer.nextNode;
+      pointer.nextNode = node;
+      node.nextNode = buf;
+    }
```
Comment (GrigoriyUstymenko) [inline src/tools/dataStructures/PriorityQueue.js:36]:
```suggestion
    }
    
    return this
```

File: src/tools/dataStructures/PriorityQueue.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      while (
+        pointer.nextNode &&
+        pointer.nextNode.priority * this.direction >= priority * this.direction
+      ) {
+        pointer = pointer.nextNode;
+      }
+
+      const buf = pointer.nextNode;
+      pointer.nextNode = node;
+      node.nextNode = buf;
```
Comment (GrigoriyUstymenko) [inline src/tools/dataStructures/PriorityQueue.js:35]: Might be a good idea to implement it this way, but Im not sure my code is working, better test it

```suggestion
    let prev = null;
      let pointer = this.head;
      while (
        pointer?.priority * this.direction >= priority * this.direction
      ) {
        prev = pointer;
        pointer = pointer.nextNode;
      }
      
      node.nextNode = pointer;
      if(prev === null) {
      	this.head = node;
        return this;
      }
      prev.nextNode = node;
    }
```

File: src/tools/dataStructures/Queue.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      this.pop();
+    }
+  }
+
+  toArray() {
+    const res = [];
+
+    for (const data of this) {
+      res.push(data);
+    }
```
Comment (GrigoriyUstymenko) [inline src/tools/dataStructures/Queue.js:51]: Why not
```suggestion
    for (const node of this) {
      res.push(node.data);
    }
```

File: src/tools/dataStructures/Tree.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  expand(vertex, ...verteces) {
+    this.insert(verteces);
+    const v1 = this.verteces.indexOf(vertex);
+    for (const v of verteces) {
+      const v2 = this.verteces.indexOf(v);
+      this.connect(v1, v2);
+    }
+  }
```
Comment (GrigoriyUstymenko) [inline src/tools/dataStructures/Tree.js:20]: Shouldnt this method add vertex as parent to each element of verteces?

**Review by GrigoriyUstymenko — CHANGES_REQUESTED** [review]
Needs some changes

File: src/tools/dataStructures/Graph.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  static isGraph(obj) {
+    return obj instanceof Graph;
+  }
+
+  #linkVerteces(verteces, matrix) {
+    for (const i in verteces) {
+      const vertex = verteces[i];
+      for (const j in matrix.rows) {
```
[the remark this answers is not in the document]
Comment (Secret333Boy) [reply src/tools/dataStructures/Graph.js:26]: Fixed

File: src/tools/dataStructures/Matrix.js [hunk 8]
```diff
@@ -31,6 +31,61 @@ class Matrix {
...
+    for (const i in this.arr) {
+      this.arr[i].splice(index, 1);
+    }
+    return this;
+  }
+
+  pushCol(col = []) {
+    for (const i in this.arr) {
+      this.arr[i].push(col[i] || null);
+    }
```
[the remark this answers is not in the document]
Comment (Secret333Boy) [reply src/tools/dataStructures/Matrix.js:55]: Fixed

File: src/tools/dataStructures/Matrix.js [hunk 9]
```diff
@@ -31,6 +31,61 @@ class Matrix {
...
+      }
+    }
+    return false;
+  }
+
+  copy() {
+    const res = [];
+    for (const i in this.arr) {
+      res.push(new Array(...this.arr[i]));
+    }
```
[the remark this answers is not in the document]
Comment (Secret333Boy) [reply src/tools/dataStructures/Matrix.js:72]: Fixed

File: src/tools/dataStructures/Matrix.js (hunk 1 above)
Comment (Secret333Boy) [reply src/tools/dataStructures/Matrix.js:166]: Fixed

File: src/tools/dataStructures/PriorityQueue.js (hunk 3 above)
Comment (Secret333Boy) [reply src/tools/dataStructures/PriorityQueue.js:36]: Fixed

File: src/tools/dataStructures/PriorityQueue.js (hunk 4 above)
Comment (Secret333Boy) [reply src/tools/dataStructures/PriorityQueue.js:35]: It works, nice solution!

Comment (Secret333Boy) [reply src/tools/dataStructures/PriorityQueue.js]: Fixed

**Comment by Secret333Boy** [comment]
Issue #4

**Pull request merged**